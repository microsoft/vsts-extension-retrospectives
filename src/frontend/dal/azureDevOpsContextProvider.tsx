import * as SDK from 'azure-devops-extension-sdk';
import React from 'react';

type SDKContextProps = {
  SDK: typeof SDK;
};

export const SDKContext = React.createContext<SDKContextProps>({} as SDKContextProps);

export const SDKProvider: React.FC = (props) => {
  const context: SDKContextProps = {
    SDK
  };

  const initSDK = async () => {
    await SDK.init();
  };

  React.useEffect(() => {
    initSDK();
  }, []);

  return (
    <SDKContext.Provider value={context}>{props.children}</SDKContext.Provider>
  );
};
