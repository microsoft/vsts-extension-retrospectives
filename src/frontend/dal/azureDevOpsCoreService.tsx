import { CoreRestClient } from 'azure-devops-extension-api/Core/CoreClient';
import { WebApiTeam } from 'azure-devops-extension-api/Core';
import { TeamMember } from 'azure-devops-extension-api/WebApi';
import { getClient } from 'azure-devops-extension-api/Common';

class AzureDevOpsCoreService {
  private _httpCoreClient: CoreRestClient;

  constructor() {
    if (!this._httpCoreClient) {
      this._httpCoreClient = getClient(CoreRestClient);
    }
  }

  public async getDefaultTeam(projectId: string): Promise<WebApiTeam> {
    return (await this._httpCoreClient.getTeams(projectId, false, 1))[0];
  }

  /**
   * Gets the teams for the current project id.
   * @param projectId The project id.
   * @param teamId The team id.
   */
  public async getTeam(projectId: string, teamId: string): Promise<WebApiTeam> {
    try {
      return await this._httpCoreClient.getTeam(projectId, teamId);
    }
    catch {
      return null;
    }
  }

  /**
   * Gets the members of teams for the current project id.
   * @param projectId The project id.
   * @param teamId The team id.
   */
  public async getMembers(projectId: string, teamId: string): Promise<TeamMember[]> {
    try {
      return await this._httpCoreClient.getTeamMembersWithExtendedProperties(projectId, teamId, 100, 0);
    }
    catch {
      return null;
    }
  }

  /**
   * Gets all the teams for the current project id.
   * @param projectId The project id.
   * @param forCurrentUserOnly If true, return teams the requesting user is a member of. If false, return teams the user can see in this project.
   */
  public async getAllTeams(projectId: string, forCurrentUserOnly: boolean): Promise<WebApiTeam[]> {
    const allTeams: WebApiTeam[] = [];

    const _httpCoreClient: CoreRestClient = getClient(CoreRestClient);

    const getTeamBatch = async (skip: number) => {
      const teamBatch: WebApiTeam[] = await _httpCoreClient.getTeams(projectId, forCurrentUserOnly, 100, skip, true);

      if (teamBatch.length > 0) {
        allTeams.push(...teamBatch);
      }

      if (teamBatch.length === 100) {
        await getTeamBatch(skip + 100);
      }
      return;
    };

    await getTeamBatch(0);

    return allTeams;
  }
}

export const azureDevOpsCoreService = new AzureDevOpsCoreService();
