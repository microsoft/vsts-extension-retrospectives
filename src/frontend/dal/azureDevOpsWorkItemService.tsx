import { WorkItemExpand, WorkItemErrorPolicy } from 'azure-devops-extension-api/WorkItemTracking/WorkItemTracking';
import { WorkItemTrackingRestClient } from 'azure-devops-extension-api/WorkItemTracking/WorkItemTrackingClient';
import { RelationshipType } from '../interfaces/workItem';

import { getClient } from 'azure-devops-extension-api/Common';
import { getProjectId } from '../utilities/servicesHelper';

class WorkItemService {
  public static readonly retrospective_type = 'Retrospective';
  public static readonly task_type = 'Task';

  private _httpClient: WorkItemTrackingRestClient;

  constructor() {
    if (!this._httpClient) {
      this._httpClient = getClient(WorkItemTrackingRestClient);
    }
  }

  public getAllFields = () => {
    return this._httpClient.getFields();
  }

  /**
   * Gets the work item states for the given work item type in the current project.
   */
  public getWorkItemStates = async (workItemType: string) => {
    const projectId = await getProjectId();

    return await this._httpClient.getWorkItemTypeStates(projectId, workItemType);
  }

  /**
   * Gets the work item types for the current project.
   */
  public getWorkItemTypesForCurrentProject = async () => {
    const projectId = await getProjectId();

    return await this._httpClient.getWorkItemTypes(projectId);
  }

  /**
   * Gets the list of work item type references for hidden work item types
   */
  public getHiddenWorkItemTypes = async () => {
    const projectId = await getProjectId();
    const hiddenWorkItemTypeCategory = await this._httpClient.getWorkItemTypeCategory(
      projectId,
      'Microsoft.HiddenCategory');

    return hiddenWorkItemTypeCategory.workItemTypes;
  }

  /**
   * Gets the work items by given ids.
   * @param ids The ids of the work items to fetch.
   */
  public async getWorkItemsByIds(ids: number[]) {
    const projectId = await getProjectId();
    const workItems = await this._httpClient.getWorkItems(ids, projectId, undefined, undefined, WorkItemExpand.All, WorkItemErrorPolicy.Omit);

    return workItems.filter(wi => wi != null);
  }

  public getReferencedByReverseItemIds(itemId: number) {
    return this.getWorkItemsByIds([itemId]).then((workItems) => {
      return workItems[0].relations.filter((relation) => relation.rel === RelationshipType.ReferencedByReverse)
        .map((relation) => Number(relation.url.split('/').pop() || ''))
        .filter((id) => id);
    });
  }

  /**
   * Gets the work items that the given work item ids have a 'Related' relationship with
   * @param itemIds
   */
  public getRelatedItemsForItemsIds(itemIds: number[]) {
    return this.getWorkItemsByIds(itemIds)
      .then((workItems) => {
        const actionItemIds = new Set<string>();
        workItems.forEach((retrospectiveItem) => {
          if (retrospectiveItem.relations) {
            retrospectiveItem.relations.filter((relation) => relation.rel === RelationshipType.Related)
              .forEach((relation) => {
                // TODO improve to just get json directly from url
                const id = relation.url.split('/').pop() || '';
                if (id) {
                  actionItemIds.add(id);
                }
              });
          }
        });
        if (actionItemIds.size) {
          return this.getWorkItemsByIds(Array.from(actionItemIds).map((e) => Number(e)));
        }
        return Promise.resolve([]);
      });
  }
}

export const workItemService = new WorkItemService();
