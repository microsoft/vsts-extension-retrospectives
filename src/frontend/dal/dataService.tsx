import { getAccessToken, getExtensionContext, getService } from 'azure-devops-extension-sdk';
import { CommonServiceIds, IExtensionDataManager, IExtensionDataService } from 'azure-devops-extension-api';
import { appInsights } from '../utilities/telemetryClient';

let extensionDataManager: IExtensionDataManager;

async function getDataService(): Promise<IExtensionDataManager> {
  if (!extensionDataManager) {
    const accessToken = await getAccessToken();
    const extensionDataService = await getService<IExtensionDataService>(CommonServiceIds.ExtensionDataService);
    extensionDataManager = await extensionDataService.getExtensionDataManager(getExtensionContext().id, accessToken);
  }

  return extensionDataManager;
}

/**
 * Read user/account scoped documents.
 */
export async function readDocuments<T>(
  collectionName: string, isPrivate?: boolean, throwCollectionDoesNotExistException?: boolean): Promise<T[]> {
  const dataService: IExtensionDataManager = await getDataService();
  let data: T[];

  try {
    data = await dataService.getDocuments(collectionName, isPrivate ? { scopeType: 'User' } : undefined);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (e: any) {
    console.error(e);
    appInsights.trackException(e);
    if (e.serverError?.typeKey === 'DocumentCollectionDoesNotExistException') {
      if (throwCollectionDoesNotExistException) {
        throw e;
      }
    }

    appInsights.trackException(e);

    data = [];
  }

  return data;
}

/**
 * Read a specific user/account scoped document.
 */
export async function readDocument<T>(collectionName: string, id: string, isPrivate?: boolean): Promise<T> {
  if (id === "emptyFeedbackItem") {
    return undefined;
  }
  const dataService: IExtensionDataManager = await getDataService();
  let data: T;
  try {
    data = await dataService.getDocument(collectionName, id, isPrivate ? { scopeType: 'User' } : undefined);
  } catch (e) {
    appInsights.trackException(e);
    console.error('An exception occurred while trying to read the document: ', e);
    data = undefined;
  }

  return data;
}

/**
 * Create user/account scoped document.
 */
export async function createDocument<T>(collectionName: string, data: T, isPrivate?: boolean): Promise<T> {
  const dataService: IExtensionDataManager = await getDataService();
  return dataService.createDocument(collectionName, data, isPrivate ? { scopeType: 'User' } : undefined);
}

/**
 * Create or Update user/account scoped document.
 */
export async function createOrUpdateDocument<T>(collectionName: string, data: T, isPrivate?: boolean): Promise<T> {
  const dataService: IExtensionDataManager = await getDataService();
  return dataService.setDocument(collectionName, data, isPrivate ? { scopeType: 'User' } : undefined);
}

/**
 * Update user/account scoped document.
 */
export async function updateDocument<T>(collectionName: string, data: T, isPrivate?: boolean): Promise<T> {
  const dataService: IExtensionDataManager = await getDataService();

  let updatedData: T;
  try {
    updatedData = await dataService.updateDocument(collectionName, data, isPrivate ? { scopeType: 'User' } : undefined);
  } catch (e) {
    appInsights.trackException(e);
    console.error('An exception occurred while trying to update the document: ', e);
    updatedData = undefined;
  }

  return updatedData;
}

/**
 * Delete user/account scoped document.
 */
export async function deleteDocument(collectionName: string, id: string, isPrivate?: boolean): Promise<void> {
  const dataService: IExtensionDataManager = await getDataService();
  return dataService.deleteDocument(collectionName, id, isPrivate ? { scopeType: 'User' } : undefined);
}

/**
 * Set user/account scoped value.
 */
export async function setValue<T>(id: string, data: T, isPrivate?: boolean): Promise<T> {
  const dataService: IExtensionDataManager = await getDataService();

  let updatedData: T;
  try {
    return dataService.setValue(id, data, isPrivate ? { scopeType: 'User' } : undefined);
  } catch (e) {
    appInsights.trackException(e);
    console.error('An exception occurred while trying to read the value: ', e);
    updatedData = undefined;
  }

  return updatedData;
}

/**
 * Get user/account scoped value.
 */
export async function getValue<T>(id: string, isPrivate?: boolean): Promise<T> {
  const dataService: IExtensionDataManager = await getDataService();

  let data: T;
  try {
    data = await dataService.getValue<T>(id, isPrivate ? { scopeType: 'User' } : undefined);
  } catch (e) {
    appInsights.trackException(e);
    console.error('An exception occurred while trying to read the value: ', e);
    data = undefined;
  }

  return data;
}
