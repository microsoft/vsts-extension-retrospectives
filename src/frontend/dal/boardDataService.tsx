import { createDocument, deleteDocument, readDocument, readDocuments, updateDocument, getValue, setValue } from './dataService';
import { IFeedbackBoardDocument, IFeedbackBoardDocumentPermissions, IFeedbackColumn, IFeedbackItemDocument } from '../interfaces/feedback';
import { WorkflowPhase } from '../interfaces/workItem';
import { getUserIdentity } from '../utilities/userIdentityHelper';
import { generateUUID } from '../utilities/random';
import { appInsights, TelemetryExceptions } from '../utilities/telemetryClient';

class BoardDataService {
  public readonly legacyPositiveColumnId: string = 'whatwentwell';
  public readonly legacyNegativeColumnId: string = 'whatdidntgowell';

  public createBoardForTeam = async (
    teamId: string,
    title: string,
    maxVotesPerUser: number,
    columns: IFeedbackColumn[],
    isIncludeTeamEffectivenessMeasurement?: boolean,
    shouldShowFeedbackAfterCollect?: boolean,
    isAnonymous?: boolean,
    startDate?: Date,
    endDate?: Date,
    permissions?: IFeedbackBoardDocumentPermissions) => {
    const boardId: string = generateUUID();
    const userIdentity = getUserIdentity();

    const board: IFeedbackBoardDocument = {
      activePhase: WorkflowPhase.Collect,
      columns,
      createdBy: userIdentity,
      createdDate: new Date(Date.now()),
      endDate,
      id: boardId,
      maxVotesPerUser,
      isIncludeTeamEffectivenessMeasurement: isIncludeTeamEffectivenessMeasurement ?? false,
      shouldShowFeedbackAfterCollect: shouldShowFeedbackAfterCollect ?? false,
      isAnonymous: isAnonymous ?? false,
      modifiedDate: new Date(Date.now()),
      startDate,
      teamId,
      title,
      boardVoteCollection: {},
      teamEffectivenessMeasurementVoteCollection: [],
      isPublic: this.isBoardPublic(permissions),
      permissions: permissions
    }

    return await createDocument<IFeedbackBoardDocument>(teamId, board);
  }

  public checkIfBoardNameIsTaken = async (teamId: string, boardName: string) => {
    const teamBoards: IFeedbackBoardDocument[] = await this.getBoardsForTeam(teamId);

    const boardNameExists =
      teamBoards.find(teamBoard => {
        return (teamBoard.title.replace(/\s+/g, ' ').trim().localeCompare(boardName.replace(/\s+/g, ' ').trim(), [],
          { sensitivity: "accent" }) === 0);
      });

    return boardNameExists ? true : false;
  }

  public getBoardsForTeam = async (teamId: string) => {
    let teamBoards: IFeedbackBoardDocument[] = [];

    try {
      teamBoards = await readDocuments<IFeedbackBoardDocument>(teamId, false, true);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (e: any) {
      appInsights.trackException(e);
      if (e.serverError?.typeKey === 'DocumentCollectionDoesNotExistException') {
        appInsights.trackTrace({ message: TelemetryExceptions.BoardsNotFoundForTeam, properties: { teamId, e } });
      }
    }

    return teamBoards;
  }

  public getBoardForTeamById = async (teamId: string, boardId: string) => {
    return await readDocument<IFeedbackBoardDocument>(teamId, boardId);
  }

  public deleteFeedbackBoard = async (teamId: string, boardId: string) => {
    // Delete all documents in this board's collection.
    const boardItems = await readDocuments<IFeedbackItemDocument>(boardId);
    if (boardItems && boardItems.length) {
      boardItems.forEach(async (item) => {
        await deleteDocument(boardId, item.id);
      });
    }

    await deleteDocument(teamId, boardId);
  }

  public archiveFeedbackBoard = async (teamId: string, boardId: string) => {
    const board: IFeedbackBoardDocument = await this.getBoardForTeamById(teamId, boardId);
    const userIdentity = getUserIdentity();

    // Check in case board was deleted by other user after option to update was selected by current user
    if (!board) {
      console.log(`Cannot archive for a non-existent feedback board. Board: ${boardId}`);
      return undefined;
    }

    board.isArchived = true;
    board.archivedDate = new Date();
    board.archivedBy = userIdentity;

    return await this.updateBoard(teamId, board);
  }

  public restoreArchivedFeedbackBoard = async (teamId: string, boardId: string) => {
    const board: IFeedbackBoardDocument = await this.getBoardForTeamById(teamId, boardId);

    // Check in case board was deleted by other user after option to update was selected by current user
    if (!board) {
      console.log(`Cannot restore for a non-existent feedback board. Board: ${boardId}`);
      return undefined;
    }

    board.isArchived = false;
    board.archivedDate = undefined;
    board.archivedBy = undefined;

    return await this.updateBoard(teamId, board);
  }

  public async saveSetting<T>(key: string, value: T): Promise<void> {
    await setValue(key, value, true); // Use 'true' to make it user-scoped
  }

  public async getSetting<T>(key: string): Promise<T> {
    const value = await getValue<T>(key, true);
    return value !== undefined ? value : (null as T); // Ensures type safety
  }

  public updateBoardMetadata = async (teamId: string, boardId: string, maxVotesPerUser: number, title: string, newColumns: IFeedbackColumn[], permissions: IFeedbackBoardDocumentPermissions): Promise<IFeedbackBoardDocument> => {
    const board: IFeedbackBoardDocument = await this.getBoardForTeamById(teamId, boardId);

    // Check in case board was deleted by other user after option to update was selected by current user
    if (!board) {
      console.log(`Cannot update title for a non-existent feedback board. Board: ${boardId}`);
      return undefined;
    }

    board.title = title;
    board.maxVotesPerUser = maxVotesPerUser;
    board.columns = newColumns;
    board.modifiedDate = new Date(Date.now());
    board.isPublic = this.isBoardPublic(permissions),
    board.permissions = permissions;

    return await this.updateBoard(teamId, board);
  }

  // Update the board document.
  private updateBoard = async (teamId: string, board: IFeedbackBoardDocument): Promise<IFeedbackBoardDocument> => {
    return await updateDocument<IFeedbackBoardDocument>(teamId, board);
  }

  private isBoardPublic = (permissions: IFeedbackBoardDocumentPermissions): boolean => {
    return permissions === undefined || (permissions.Teams.length === 0 && permissions.Members.length === 0);
  }
}

export default new BoardDataService();
