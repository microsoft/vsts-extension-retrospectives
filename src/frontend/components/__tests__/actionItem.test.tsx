import React from "react";
import { shallow, mount } from 'enzyme';
import toJson from "enzyme-to-json";
import { ActionItem, ActionItemProps } from "../actionItem";
import { DocumentCardActions, DocumentCardTitle, DocumentCardPreview, IDocumentCardPreviewProps } from 'office-ui-fabric-react/lib/DocumentCard';
import Dialog from 'office-ui-fabric-react/lib/Dialog';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { itemDataService } from '../../dal/itemDataService';
import { workItemService } from '../../dal/azureDevOpsWorkItemService';
import { IFeedbackItemDocument } from '../../interfaces/feedback';
import { WorkItemType, WorkItem } from 'azure-devops-extension-api/WorkItemTracking/WorkItemTracking';

const mockOnUpdateActionItem = jest.fn(() => { });

const defaultTestProps: ActionItemProps = {
  feedbackItemId: "101",
  boardId: "Test Board Id",
  nonHiddenWorkItemTypes: [],
  allWorkItemTypes: [
    {
      color: "red",
      description: "Test description",
      icon: { id: "1", url: "testUrl" },
      isDisabled: false,
      name: "Test Work Item Type",
      referenceName: "Test Reference Name",
      fieldInstances: [],
      fields: [],
      states: [],
      transitions: {},
      xmlForm: "Test xmlForm",
      _links: {},
      url: "Test url"
    }
  ],
  onUpdateActionItem: mockOnUpdateActionItem,
  actionItem: {
    _links: {},
    url: "Test url",
    id: 1,
    commentVersionRef: { commentId: 1, version: 1, url: "Test url", createdInRevision: 1, isDeleted: false, text: "Test text" },
    relations: [],
    rev: 1,
    fields: {
      "System.Title": "Test Title",
      "System.WorkItemType": "Test Work Item Type",
    },
  },
  areActionIconsHidden: false,
  shouldFocus: false
};

describe("Action Item component", () => {
  it("renders correctly when there are no action items.", () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correctly when action items exist and areActionIconsHidden is true", () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} areActionIconsHidden={true} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correctly when action items exist and shouldFocus is true", () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} shouldFocus={true} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correctly when action items exist and areActionIconsHidden is false", () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} areActionIconsHidden={false} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correctly when action items exist and shouldFocus is false", () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} shouldFocus={false} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("focuses the openWorkItemButton when shouldFocus is true", () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} shouldFocus={true} />);
    const instance = wrapper.instance() as ActionItem;
    const mockFocus = jest.fn();
    instance.openWorkItemButton = { focus: mockFocus } as unknown as HTMLButtonElement;
    instance.componentDidMount();
    expect(mockFocus).toHaveBeenCalled();
  });
});

describe('Behavioral tests for ActionItem', () => {
  it('does not render actions when areActionIconsHidden is true', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} areActionIconsHidden={true} />);
    expect(wrapper.find(DocumentCardActions)).toHaveLength(0);
  });

  it('renders actions when areActionIconsHidden is false', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} areActionIconsHidden={false} />);
    expect(wrapper.find(DocumentCardActions)).toHaveLength(1);
  });

  it('toggles unlink confirmation dialog visibility based on state', () => {
    const wrapper = mount(<ActionItem {...defaultTestProps} />);
    // initially dialog is hidden
    expect(wrapper.find(Dialog)).toHaveLength(0);
    // show dialog
    wrapper.setState({ isUnlinkWorkItemConfirmationDialogHidden: false });
    wrapper.update();
    expect(wrapper.find(Dialog)).toHaveLength(1);
    // hide dialog
    wrapper.setState({ isUnlinkWorkItemConfirmationDialogHidden: true });
    wrapper.update();
    expect(wrapper.find(Dialog)).toHaveLength(0);
  });

  it('truncates titles longer than 25 characters', () => {
    const longTitle = 'A'.repeat(30);
    const modifiedProps = {
      ...defaultTestProps,
      actionItem: {
        ...defaultTestProps.actionItem,
        fields: { ...defaultTestProps.actionItem.fields, 'System.Title': longTitle }
      }
    };
    const wrapper = shallow(<ActionItem {...modifiedProps} />);
    const titleText = wrapper.find(DocumentCardTitle).prop('title');
    expect(titleText.length).toBe(28); // 25 chars + '...'
    expect(titleText.endsWith('...')).toBe(true);
  });
});

describe('UI-level integration tests for ActionItem', () => {
  it('renders the correct icon in DocumentCardPreview', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const preview = wrapper.find(DocumentCardPreview);
    const previewImages = preview.prop('previewImages') as IDocumentCardPreviewProps['previewImages'];
    expect(previewImages![0].previewIconProps.imageProps.src).toBe('testUrl');
  });

  it('applies resolved-border-right class when work item state is Completed', () => {
    const modifiedProps = {
      ...defaultTestProps,
      actionItem: {
        ...defaultTestProps.actionItem,
        fields: {
          ...defaultTestProps.actionItem.fields,
          'System.State': 'Completed'
        }
      },
      allWorkItemTypes: [
        { ...defaultTestProps.allWorkItemTypes[0], states: [{ name: 'Completed', category: 'Completed', color: 'blue' }] }
      ]
    };
    const wrapper = shallow(<ActionItem {...modifiedProps} />);
    expect(wrapper.find('.related-task-sub-card').hasClass('resolved-border-right')).toBe(true);
  });

  it('calls onUpdateActionItem when confirming unlink', async () => {
    jest.spyOn(itemDataService, 'removeAssociatedActionItem').mockResolvedValue({} as IFeedbackItemDocument);
    const wrapper = mount(<ActionItem {...defaultTestProps} />);
    // show dialog by setting state
    wrapper.setState({ isUnlinkWorkItemConfirmationDialogHidden: false });
    wrapper.update();
    const removeBtn = wrapper.find(PrimaryButton).filterWhere(b => b.prop('text') === 'Remove');
    removeBtn.simulate('click', { stopPropagation: () => { }, preventDefault: () => { } });
    // Wait for promises to resolve (flush microtasks)
    await Promise.resolve();
    await Promise.resolve();
    expect(defaultTestProps.onUpdateActionItem).toHaveBeenCalled();
  });
});

describe('Accessibility and edge case tests for ActionItem', () => {
  it('opens work item form on Enter key press', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const detailsDiv = wrapper.find('.ms-DocumentCard-details');
    const instance = wrapper.instance() as ActionItem;
    const spy = jest.fn();
    // @ts-expect-error: override private for test
    instance.onActionItemClick = spy;
    detailsDiv.simulate('keyPress', { key: 'Enter', stopPropagation: () => { } });
    expect(spy).toHaveBeenCalledWith(defaultTestProps.actionItem.id);
  });

  it('hides dialog when cancel button is clicked', () => {
    const wrapper = mount(<ActionItem {...defaultTestProps} />);
    wrapper.setState({ isUnlinkWorkItemConfirmationDialogHidden: false });
    wrapper.update();
    const cancelBtn = wrapper.find(DefaultButton).filterWhere(b => b.prop('text') === 'Cancel');
    cancelBtn.simulate('click', { stopPropagation: () => { }, preventDefault: () => { } });
    expect(wrapper.find(Dialog)).toHaveLength(0);
  });

  it('renders without icon if work item type is not found', () => {
    // Provide a stub workItemType with minimal properties to avoid undefined errors
    const props = { ...defaultTestProps, allWorkItemTypes: [{ name: '', icon: { url: '' } }] as WorkItemType[] };
    const wrapper = shallow(<ActionItem {...props} />);
    const preview = wrapper.find(DocumentCardPreview);
    expect(preview.exists()).toBe(true);
  });

  it('does not throw if work item type is missing (allWorkItemTypes empty)', () => {
    const props = { ...defaultTestProps, allWorkItemTypes: [] as WorkItemType[] };
    expect(() => shallow(<ActionItem {...props} />)).not.toThrow();
  });
});

describe('Private method coverage tests', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('showUnlinkWorkItemConfirmationDialog sets state correctly', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;

    // @ts-expect-error: accessing private method for test
    instance.showUnlinkWorkItemConfirmationDialog();

    expect(wrapper.state('isUnlinkWorkItemConfirmationDialogHidden')).toBe(false);
  });

  it('hideUnlinkWorkItemConfirmationDialog sets state correctly', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    wrapper.setState({ isUnlinkWorkItemConfirmationDialogHidden: false });
    const instance = wrapper.instance() as ActionItem;

    // @ts-expect-error: accessing private method for test
    instance.hideUnlinkWorkItemConfirmationDialog();

    expect(wrapper.state('isUnlinkWorkItemConfirmationDialogHidden')).toBe(true);
  });

  it('onConfirmUnlinkWorkItem calls onUnlinkWorkItemClick and hides dialog', async () => {
    jest.spyOn(itemDataService, 'removeAssociatedActionItem').mockResolvedValue({} as IFeedbackItemDocument);
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;

    wrapper.setState({ isUnlinkWorkItemConfirmationDialogHidden: false });

    // @ts-expect-error: accessing private method for test
    await instance.onConfirmUnlinkWorkItem(defaultTestProps.actionItem.id);

    expect(defaultTestProps.onUpdateActionItem).toHaveBeenCalled();
    expect(wrapper.state('isUnlinkWorkItemConfirmationDialogHidden')).toBe(true);
  });

  it('updateLinkedItem updates state when linkedWorkItem matches ID', async () => {
    const mockUpdatedWorkItem = {
      id: 123,
      fields: { 'System.Title': 'Updated' },
      commentVersionRef: { commentId: 1, version: 1, url: "test", createdInRevision: 1, isDeleted: false, text: "test" },
      relations: [],
      rev: 1,
      _links: {},
      url: "test"
    } as WorkItem;
    jest.spyOn(workItemService, 'getWorkItemsByIds').mockResolvedValue([mockUpdatedWorkItem]);

    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;

    wrapper.setState({ linkedWorkItem: { id: 123 } });

    // @ts-expect-error: accessing private method for test
    await instance.updateLinkedItem(123);

    expect(wrapper.state('linkedWorkItem')).toEqual(mockUpdatedWorkItem);
  });

  it('updateLinkedItem does not update when linkedWorkItem ID does not match', async () => {
    const mockUpdatedWorkItem = {
      id: 456,
      fields: { 'System.Title': 'Updated' },
      commentVersionRef: { commentId: 1, version: 1, url: "test", createdInRevision: 1, isDeleted: false, text: "test" },
      relations: [],
      rev: 1,
      _links: {},
      url: "test"
    } as WorkItem;
    jest.spyOn(workItemService, 'getWorkItemsByIds').mockResolvedValue([mockUpdatedWorkItem]);

    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;
    const originalLinkedItem = { id: 123 };

    wrapper.setState({ linkedWorkItem: originalLinkedItem });

    // @ts-expect-error: accessing private method for test
    await instance.updateLinkedItem(456);

    expect(wrapper.state('linkedWorkItem')).toEqual(originalLinkedItem);
  });

  it('handleKeyPressSelectorButton calls showUnlinkWorkItem on Enter key', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;
    const spy = jest.fn();

    // @ts-expect-error: override private method for test
    instance.showUnlinkWorkItem = spy;

    const event = { key: 'Enter', preventDefault: jest.fn(), stopPropagation: jest.fn() };

    // @ts-expect-error: accessing private method for test
    instance.handleKeyPressSelectorButton(event);

    expect(spy).toHaveBeenCalledWith(event);
  });

  it('handleKeyPressSelectorButton does not call showUnlinkWorkItem on other keys', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;
    const spy = jest.fn();

    // @ts-expect-error: override private method for test
    instance.showUnlinkWorkItem = spy;

    const event = { key: 'Space', preventDefault: jest.fn(), stopPropagation: jest.fn() };

    // @ts-expect-error: accessing private method for test
    instance.handleKeyPressSelectorButton(event);

    expect(spy).not.toHaveBeenCalled();
  });

  it('showUnlinkWorkItem prevents default and stops propagation', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;

    const event = {
      preventDefault: jest.fn(),
      stopPropagation: jest.fn()
    };

    // @ts-expect-error: accessing private method for test
    instance.showUnlinkWorkItem(event);

    expect(event.preventDefault).toHaveBeenCalled();
    expect(event.stopPropagation).toHaveBeenCalled();
    expect(wrapper.state('isUnlinkWorkItemConfirmationDialogHidden')).toBe(false);
  });

  it('showWorkItemForm stops propagation and calls onActionItemClick', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;
    const spy = jest.fn();

    // @ts-expect-error: override private method for test
    instance.onActionItemClick = spy;

    const event = { stopPropagation: jest.fn() };

    // @ts-expect-error: accessing private method for test
    instance.showWorkItemForm(event);

    expect(event.stopPropagation).toHaveBeenCalled();
    expect(spy).toHaveBeenCalledWith(defaultTestProps.actionItem.id);
  });

  it('showWorkItemForm handles null/undefined event', () => {
    const wrapper = shallow(<ActionItem {...defaultTestProps} />);
    const instance = wrapper.instance() as ActionItem;
    const spy = jest.fn();

    // @ts-expect-error: override private method for test
    instance.onActionItemClick = spy;

    // @ts-expect-error: accessing private method for test
    instance.showWorkItemForm(null);

    expect(spy).toHaveBeenCalledWith(defaultTestProps.actionItem.id);
  });
});
