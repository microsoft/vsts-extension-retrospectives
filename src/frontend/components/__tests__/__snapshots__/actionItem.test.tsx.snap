// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Item component renders correctly when there are no action items. 1`] = `
<StyledDocumentCardBase
  className="related-task-sub-card "
  key="1card"
  type={1}
>
  <StyledDocumentCardPreviewBase
    key="1preview"
    previewImages={
      Array [
        Object {
          "previewIconContainerClass": "work-item-type-icon-container",
          "previewIconProps": Object {
            "ariaLabel": "icon for work item type Test Work Item Type",
            "imageProps": Object {
              "alt": "icon for work item type Test Work Item Type",
              "src": "testUrl",
            },
          },
          "width": 36,
        },
      ]
    }
  />
  <div
    aria-label="Test Work Item Type Test Title, click to open work item"
    className="ms-DocumentCard-details"
    key="1details"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <StyledDocumentCardTitleBase
      key="1title"
      shouldTruncate={true}
      title="Test Title"
    />
  </div>
  <StyledDocumentCardActionsBase
    actions={
      Array [
        Object {
          "ariaLabel": "Remove link to work item button",
          "iconProps": Object {
            "iconName": "RemoveLink",
          },
          "onClick": [Function],
          "onKeyPress": [Function],
          "title": "Remove link to work item",
        },
      ]
    }
  />
</StyledDocumentCardBase>
`;
